### PR TITLE
validation

### PR DESCRIPTION
**Pull request by nikk-nikaznan**
Adding validation set for the training.

- make changes on evalution_model to reflect the validation set
- now the `test` is using lightning as well with mlflowlogger
- moved the prep functions for images_dirs and annotations to `detection_utils` so both training and test can use them. (let me know if it's okay or not)

### CODE REVIEW

File: crabs/detection_tracking/evaluate_model.py [hunk 1]
```diff
@@ -192,6 +194,9 @@ def main(args) -> None:
         action="store_true",
         help=("Save predicted frames with bboxes."),
     )
+    return parser.parse_args(args)
+
 
-    args = parser.parse_args()
+if __name__ == "__main__":
+    args = evaluate_parse_args(sys.argv[1:])
```
Comment (sfmig) [inline crabs/detection_tracking/evaluate_model.py:201]: nice! ✨

**Review by sfmig — APPROVED** [review]
thanks Nik! 
It took a bit longer cause I was trying to see if we could make the evaluate script more lightning-y. Specifically I was trying to bypass the manual loading of a trained model, by using the checkpoint input argument of `trainer.test` instead... I reduced a bit the number of lines but didn't crack it yet.

I noticed some types for inputs and returns are missing, if you could add those that would be great.

I made some changes to the eval script to make the CLI argument parsing consistent with the training script one. I will push it if that is alright, let me know what you think. If you can still check the comments I left in the original code to review,  I think that would help understand the changes to the eval script.

File: crabs/detection_tracking/evaluate.py [hunk 2]
```diff
@@ -11,7 +12,7 @@
 
 
 def save_images_with_boxes(
```
[the remark this answers is not in the document]
Comment (sfmig) [reply crabs/detection_tracking/evaluate.py:14]: also: should we set the device to match the device where the trained model is instead? (this could be another PR but we could make an issue for now)

File: crabs/detection_tracking/models.py [hunk 3]
```diff
@@ -67,17 +77,89 @@ def on_train_epoch_end(self):
...
             {"train_loss": avg_loss}, step=self.current_epoch
         )
-
-        # Reset the training_step_outputs for the next epoch
         self.training_step_outputs = {
             "total_training_loss": 0.0,
             "num_batches": 0,
         }
 
+    def on_validation_epoch_end(self):
```
[the remark this answers is not in the document]
Comment (nikk-nikaznan) [reply crabs/detection_tracking/models.py]: This is tricky. The reason why I ended up need to use precision and not loss to train optuna is because the output during `model(input)` for training which gives loss not giving loss during eval. It returns the prediction output. It's possible but quite a pain. So maybe need to do next TI

File: crabs/detection_tracking/evaluate_model.py [hunk 4]
```diff
@@ -166,13 +168,13 @@ def main(args) -> None:
...
         type=float,
-        default=0.5,
+        default=0.1,
         help="threshold for confidence score",
     )
     parser.add_argument(
         "--ious_threshold",
         type=float,
-        default=0.5,
+        default=0.1,
```
[the remark this answers is not in the document]
Comment (nikk-nikaznan) [reply crabs/detection_tracking/evaluate_model.py:177]: During the recent tracking, we use 0.1 mainly to make the tracking better. especially a lot of the boxes are small. Afterall it's a parameter

File: crabs/detection_tracking/evaluate.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nikk-nikaznan) [reply crabs/detection_tracking/evaluate.py:14]: I saw there is a PR for that, we can move it for that PR

File: crabs/detection_tracking/evaluate.py [hunk 5]
```diff
@@ -29,6 +30,14 @@ def save_images_with_boxes(
     ----------
         None
     """
+    device = (
+        torch.device("cuda")
+        if torch.cuda.is_available()
+        else torch.device("cpu")
+    )
+    trained_model.eval()
```
[the remark this answers is not in the document]
Comment (nikk-nikaznan) [reply crabs/detection_tracking/evaluate.py:38]: I tried to make it works with lightning, but it is a bit trickier to make it works with the function we already have. And as it is optional which we probably never used much (not using at all this TI), I will say leave it for now.

File: crabs/detection_tracking/models.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sfmig) [reply crabs/detection_tracking/models.py]: aah that is annoying! ok let's keep an eye on it, and consider attacking it next TI

File: crabs/detection_tracking/evaluate_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sfmig) [reply crabs/detection_tracking/evaluate_model.py:177]: yeah I thought 0.5 was fairly standard so that's why I asked. Fair enough!

File: crabs/detection_tracking/models.py [hunk 6]
```diff
@@ -1,83 +1,216 @@
...
+        """
         return self.model(x)
 
-    def training_step(self, batch, batch_idx):
-        images, targets = batch
-        loss_dict = self.model(images, targets)
-        total_loss = sum(loss for loss in loss_dict.values())
+    def accumulate_metrics(
+        self,
+        outputs: List[Dict[str, Union[float, int]]],
```
Comment (sfmig) [inline crabs/detection_tracking/models.py:85]: when we use this function, isn't outputs always a 1-element list? should we then simplify this function?
